lib/Service/ShareService.php

```diff
@@ -0,0 +1,112 @@
+<?php
+/**
+ * Nextcloud - News
+ *
+ * This file is licensed under the Affero General Public License version 3 or
+ * later. See the COPYING file.
+ *
+ * @author    Marco Nassabain <marco.nassabain@hotmail.com>
+ */
+
+namespace OCA\News\Service;
+
+use \OCA\News\Db\Item;
+use \OCA\News\Db\Feed;
+
+use \Psr\Log\LoggerInterface;
+
+use OCA\News\Service\Exceptions\ServiceNotFoundException;
+use OCP\AppFramework\Db\DoesNotExistException;
+
+/**
+ * Class ShareService
+ *
+ * @package OCA\News\Service
+ */
+class ShareService
+{
+    /**
+     * Items service.
+     *
+     * @var ItemServiceV2
+     */
+    protected $itemService;
+
+    /**
+     * Feeds service.
+     *
+     * @var FeedServiceV2
+     */
+    protected $feedService;
+    
+    /**
+     * @var LoggerInterface
+     */
+    protected $logger;
+
+    /**
+     * ShareService constructor
+     *
+     * @param FeedServiceV2   $feedService Service for feeds
+     * @param ItemServiceV2   $itemService Service to manage items
+     * @param LoggerInterface $logger      Logger
+     */
+    public function __construct(
+        FeedServiceV2 $feedService,
+        ItemServiceV2 $itemService,
+        LoggerInterface $logger
+    ) {
+        $this->itemService = $itemService;
+        $this->feedService = $feedService;
+        $this->logger      = $logger;
+    }
+
+    /**
+     * Share an item with a user
+     *
+     * @param string $userId           ID of user sharing the item
+     * @param int    $itemId           Item ID
+     * @param string $shareRecipientId ID of user to share with
+     *
+     * Sharing by copying - the item is duplicated, and the 'sharedBy'
+     * field is filled accordingly.
+     * The item is then placed in a dummy feed reserved for items
+     * shared with the user
+     *
+     * @return Item Shared item
+     * @throws ServiceNotFoundException|ServiceConflictException
+     */
+    public function shareItemWithUser(string $userId, int $itemId, string $shareRecipientId)
+    {
+        // Find item to share
+        $item = $this->itemService->find($userId, $itemId);
+
+        // Duplicate item & initialize fields
+        $sharedItem = clone $item;
+        $sharedItem->setId(null);
```

Comment:
MySQL/MariaDB and SQLite would use the auto-increment value if `NULL` is provided, but I don't know the doctrine handling of it.
And the ID of the cloned item has to be reset for sure.

Reply:
It wasn't valid at first, I had to implement a __clone() method in Item that, in addition to cloning the element, marks all the feeds as updated. This was necessary because the insert() function only inserts columns that are marked as updated.
I didn't know whether I should put the setId(null) in that function or not, but we think it's better outside since a clone should arguably have the same id as the cloned item.

Anyways, I've tested the api call and it works, the item is inserted with a new id